translator/GCP_cloud_function/cloud_function/main.py

```diff
@@ -1,35 +1,53 @@
-# {'bucket': 'icone_files', 'contentType': 'image/jpeg', 'crc32c': 'ZO4N6A==', 'etag': 'CNu+up7tnu4CEAE=', 
-# 'generation': '1610745138552667', 'id': 'icone_files/61862931_1039122999618882_7010636350885986304_o_1039122996285549.JPEG/1610745138552667', 
-# 'kind': 'storage#object', 'md5Hash': 'yOlwCtFi0L/3GPe23rYGbw==', 
-# 'mediaLink': 'https://www.googleapis.com/download/storage/v1/b/icone_files/o/61862931_1039122999618882_7010636350885986304_o_1039122996285549.JPEG?generation=1610745138552667&alt=media', 
-# 'metageneration': '1', 'name': '61862931_1039122999618882_7010636350885986304_o_1039122996285549.JPEG',
-#  'selfLink': 'https://www.googleapis.com/storage/v1/b/icone_files/o/61862931_1039122999618882_7010636350885986304_o_1039122996285549.JPEG',
-#   'size': '447421', 'storageClass': 'STANDARD', 'timeCreated': '2021-01-15T21:12:18.552Z', 'timeStorageClassUpdated': '2021-01-15T21:12:18.552Z', 
-#   'updated': '2021-01-15T21:12:18.552Z'}
-
 import icone_translator
 from gcloud import storage
 import json
 import xmltodict
+import shutil
+import urllib.request as request
+from contextlib import closing
+from google.cloud import pubsub_v1
+
+
+server = 'iconetraffic.com'
+port = '42663'
+user = 'cdot'
+password = 'icone_cdot'
+filepath = 'incidents.xml'
+
+
+ftpString = 'ftp://{0}:{1}@{2}:{3}/{4}'
+ftpUrl = ftpString.format(user, password, server, port, filepath)
+
+
+def hello_pubsub(event, context):
+
+
+  icone_data=get_ftp_file(ftpUrl)
+  icone_obj=parse_xml(icone_data)
+  info=icone_translator.initialize_info()
+
+  wzdx_obj=icone_translator.wzdx_creator(icone_obj,info)
+  wzdx_schema=get_wzdx_schema('wzdx_schema.json')
+  validation_result=icone_translator.validate_wzdx(wzdx_obj,wzdx_schema)
+  # if not validation_result:
+  #   print('validation failed! see previous messages for error messages')
```

Comment:
Dead code, can we remove?

Reply:
Actually, maybe this should be uncommented? It looks like we get a "validation_result" and then proceed without checking it. We want to be ensuring only valid data gets through.

Reply:
We removed the commented code.
This will throw an exception if validation fails, so there is no reason to check the result. We removed the variable assignment to reflect that.

Reply:
What happens when the `validate_wzdx` function throws an exception? Does the whole app crash? Do we get any logging to help troubleshoot what went wrong? We need to be checking the validation and logging any issues so we can figure out what went wrong in the future.

Reply:
The cloud function will crash and stop execution. The ValidationError thrown by the validate_wzdx function has all of the information about what field failed validation and why, making it (usually) easy to diagnose the issue. It looks like we can set up alerts (email or text) for if the function fails, that would be a good thing to do.

Reply:
We really should be handling any exceptions that would cause the entire app to crash. The exception should be caught, logged to output, and then allow the app to exit gracefully.

Reply:
The way I am thinking about it, I want the app to crash and throw an error, so that we can easily tell when it fails. If we simply log an error message, then it is much less obvious that something went wrong. 
My thinking is that if the WZDx message validation fails, then the WZDx message is invalid and an uncaught error has occurred somewhere, meaning that the function has failed

Reply:
If we decide that we still want to publish an invalid WZDx message, than we should definitely be handling the error. Otherwise, we need to be sounding alarm bells because something broke, and the easiest way I can see to do that is to crash the function app, leaving traceback info in the cloud function logs.

Reply:
If we want to raise alarm bells for an invalid message (which makes sense as this is an odd issue), we should add in some notification logic to either send out an email or fire off some other trigger. The app itself should never be designed to fail though. With current logic allowing the app to crash, is there some notification we have set up to alert on cloud function fail, or would the GCP manager see this on a dashboard view or another location?

Reply:
@abinashkonersman Can you write up a story for improving the error reporting (and not crashing the app)? this link seems relevant: https://cloud.google.com/functions/docs/monitoring/error-reporting

Reply:
https://cdot-its.atlassian.net/browse/CV-64?atlOrigin=eyJpIjoiMDA3OTAwZDFlYzNkNDJiZmFkMTQ5ZWJkYjBkMmIyNDQiLCJwIjoiaiJ9 Seems to have been written to handle this. Closing.